<a href="https://colab.research.google.com/github/VitalyGladyshev/LangChain-Experiments/blob/main/YandexGPT_%D1%87%D0%B5%D1%80%D0%B5%D0%B7_LangCHain.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Проба YandexGPT через LangCHain

In [ ]:
Сimport pickle

In [ ]:
with open("headers", "rb") as fp:
    headers = pickle.load(fp)

In [ ]:
with open("chapters", "rb") as fp:
    chapters = pickle.load(fp)

In [ ]:
type(headers)

list

In [ ]:
headers[0]

'Назначение программного модуля CAD'

In [ ]:
type(chapters)

dict

In [ ]:
chapters[headers[0]]

'Программный модуль CAD разработан на базе интегрированной инженерной программной платформы и программно-математического ядра трехмерного моделирования «RGK».Функциональное назначение программного модуля CAD:создание и редактирование геометрической модели объектов проектирования, включая 3D- и 2D-представления;задание атрибутивной информации, связанной с точностными, размерными, технологическими параметрами объектов проектирования (PMI);оформление графической документации в соответствии с требованиями стандартов (ЕСКД);визуализация объектов проектирования с различными параметрами отображения.'

## Yandex

In [ ]:
# %pip install yandexcloud

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 47.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 56.3 MB/s eta 0:00:00
  Attempting uninstall: cryptography
    Found existing installation: cryptography 43.0.3
    Uninstalling cryptography-43.0.3:
      Successfully uninstalled cryptography-43.0.3


In [ ]:
import os
import getpass

In [ ]:
def _set_env(var: str):
    if not os.environ.get(var):
        os.environ[var] = getpass.getpass(f"{var}: ")

In [ ]:
_set_env("YC_IAM_TOKEN")

YC_IAM_TOKEN: ··········


In [ ]:
_set_env("YC_API_KEY")

YC_API_KEY: ··········


In [ ]:
_set_env("YC_FOLDER_ID")

YC_FOLDER_ID: ··········


### LLMs YandexGPT example

In [ ]:
%pip list

Package                            Version
---------------------------------- --------------------
absl-py                            1.4.0
accelerate                         1.1.1
aiohappyeyeballs                   2.4.3
aiohttp                            3.11.1
aiosignal                          1.3.1
alabaster                          1.0.0
albucore                           0.0.19
albumentations                     1.4.20
altair                             4.2.2
annotated-types                    0.7.0
anyio                              3.7.1
argon2-cffi                        23.1.0
argon2-cffi-bindings               21.2.0
array_record                       0.5.1
arviz                              0.20.0
astropy                            6.1.6
astropy-iers-data                  0.2024.11.11.0.32.38
astunparse                         1.6.3
async-timeout                      4.0.3
atpublic                           4.1.0
attrs                              24.2.0
audioread         

In [ ]:
%pip install --upgrade langchain

In [ ]:
%pip install langchain-community

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 67.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 40.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.5/49.5 kB 4.1 MB/s eta 0:00:00
  Attempting uninstall: SQLAlchemy
    Found existing installation: SQLAlchemy 2.0.36
    Uninstalling SQLAlchemy-2.0.36:
      Successfully uninstalled SQLAlchemy-2.0.36


In [ ]:
from langchain.chains import LLMChain
from langchain_community.llms import YandexGPT
from langchain_core.prompts import PromptTemplate

/usr/local/lib/python3.10/dist-packages/pydantic/_internal/_fields.py:132: UserWarning: Field "model_uri" in _BaseYandexGPT has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/pydantic/_internal/_fields.py:132: UserWarning: Field "model_name" in _BaseYandexGPT has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/pydantic/_internal/_fields.py:132: UserWarning: Field "model_version" in _BaseYandexGPT has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(


In [ ]:
template = "What is the capital of {country}?"
prompt = PromptTemplate.from_template(template)

In [ ]:
yandex_gpt = YandexGPT(folder_id="b1gt3tgpol6rntep2m65")
# yandex_gpt = YandexGPT(iam_token="y0_A...", folder_id="b1gt3tgpol6rntep2m65")

In [ ]:
llm_chain = LLMChain(prompt=prompt, llm=yandex_gpt)

In [ ]:
country = "Russia"

llm_chain.invoke(country)

{'country': 'Russia', 'text': 'The capital of Russia is **Moscow**.'}

In [ ]:
yandex_gpt("Что ты умеешь?")

<ipython-input-29-62a5de514b82>:1: LangChainDeprecationWarning: The method `BaseLLM.__call__` was deprecated in langchain-core 0.1.7 and will be removed in 1.0. Use :meth:`~invoke` instead.
  yandex_gpt("Что ты умеешь?")


'Я — модель искусственного интеллекта YandexGPT, которая умеет решать различные задачи и отвечать на вопросы. Я могу:\n\n* **Искать информацию в интернете:** я могу помочь найти ответы на ваши вопросы, предоставляя ссылки на авторитетные источники.\n\n* **Писать тексты:** я умею писать разные виды текстов — от простых до сложных. Например, я могу написать новость, статью, письмо, сочинение или даже книгу.\n\n* **Переводить текст:** я могу перевести текст с одного языка на другой, например, с английского на русский, с русского на английский, с английского на китайский и так далее.\n\n* **Решать математические задачи:** я могу решить простые и сложные математические задачи, объяснить решение или проверить его правильность.\n\n* **Помогать с изучением языков:** я могу предложить упражнения для практики грамматики, лексики и произношения, а также помочь с переводом текстов.\n\n* **Генерировать идеи:** я могу придумать новые идеи для проектов, бизнеса, креативных задач или просто для развле

Я — модель искусственного интеллекта YandexGPT, которая умеет решать различные задачи и отвечать на вопросы. Я могу:

* **Искать информацию в интернете:** я могу помочь найти ответы на ваши вопросы, предоставляя ссылки на авторитетные источники.

* **Писать тексты:** я умею писать разные виды текстов — от простых до сложных. Например, я могу написать новость, статью, письмо, сочинение или даже книгу.

* **Переводить текст:** я могу перевести текст с одного языка на другой, например, с английского на русский, с русского на английский, с английского на китайский и так далее.

* **Решать математические задачи:** я могу решить простые и сложные математические задачи, объяснить решение или проверить его правильность.

* **Помогать с изучением языков:** я могу предложить упражнения для практики грамматики, лексики и произношения, а также помочь с переводом текстов.

* **Генерировать идеи:** я могу придумать новые идеи для проектов, бизнеса, креативных задач или просто для развлечения.

* **Давать советы:** я могу дать совет по различным вопросам, связанным с вашей жизнью, работой, учёбой или другими областями.

Пожалуйста, учтите, что я не могу предоставить конкретные советы по юридическим, медицинским или финансовым вопросам. В таких ситуациях рекомендуется обратиться к специалисту.

In [ ]:
y_gpt = YandexGPT()

In [ ]:
y_gpt("Чей Крым?")

'В интернете есть много сайтов с информацией на эту тему. [Посмотрите, что нашлось в поиске](https://ya.ru)'

In [ ]:
y_gpt("На территории какого государства находится Крым?")

'В интернете есть много сайтов с информацией на эту тему. [Посмотрите, что нашлось в поиске](https://ya.ru)'

In [ ]:
y_gpt("Расскажи про Москву")

'**Москва** — столица и крупнейший город России, один из самых крупных городов Европы. Она расположена в центре европейской части страны и является важным политическим, экономическим, культурным и туристическим центром.\n\n**История:** Москва была основана в 1147 году князем Юрием Долгоруким и с тех пор стала одним из важнейших центров Руси. В разные периоды она была столицей различных государств и имела большое значение в истории России. В Москве расположены многие исторические и культурные памятники, которые привлекают туристов со всего мира.\n\n**География:** Москва находится в центральной части европейской России, на берегу реки Москвы. Город занимает площадь около 2,5 тысяч квадратных километров и насчитывает более 12 миллионов жителей. Москва граничит с Московской областью и имеет развитую транспортную инфраструктуру.\n\n**Достопримечательности:** В Москве множество достопримечательностей, среди которых можно выделить Кремль, Красную площадь, храм Христа Спасителя, Большой театр,

**Москва** — столица и крупнейший город России, один из самых крупных городов Европы. Она расположена в центре европейской части страны и является важным политическим, экономическим, культурным и туристическим центром.

**История:** Москва была основана в 1147 году князем Юрием Долгоруким и с тех пор стала одним из важнейших центров Руси. В разные периоды она была столицей различных государств и имела большое значение в истории России. В Москве расположены многие исторические и культурные памятники, которые привлекают туристов со всего мира.

**География:** Москва находится в центральной части европейской России, на берегу реки Москвы. Город занимает площадь около 2,5 тысяч квадратных километров и насчитывает более 12 миллионов жителей. Москва граничит с Московской областью и имеет развитую транспортную инфраструктуру.

**Достопримечательности:** В Москве множество достопримечательностей, среди которых можно выделить Кремль, Красную площадь, храм Христа Спасителя, Большой театр, Новодевичий монастырь, Арбат и другие. Также в городе расположено множество музеев, театров, парков и других мест, которые можно посетить.

**Культура:** Москва является центром российской культуры. Здесь находятся многие известные театры, концертные залы, музеи и выставочные центры. В городе проводятся различные фестивали, концерты и другие культурные мероприятия.

**Экономика:** Москва — один из крупнейших экономических центров России. Здесь расположены многие крупные компании, банки и организации. Город имеет развитую промышленность, торговлю и сферу услуг.

**Образование:** В Москве расположено множество высших учебных заведений, включая ведущие университеты и институты страны. Также в городе есть школы, колледжи и другие образовательные учреждения.

**Транспорт:** Москва имеет развитую систему общественного транспорта, которая включает метро, автобусы, троллейбусы, трамваи и такси. Также в городе расположены международные аэропорты, железнодорожные станции и другие транспортные узлы.

Москва — это город с богатой историей, культурой и возможностями. Он привлекает туристов своими достопримечательностями, развитой инфраструктурой и возможностями для отдыха и развлечений.

In [ ]:
y_gpt_lt = YandexGPT(model_uri="gpt://b1gt3tgpol6rntep2m65/yandexgpt/latest")

In [ ]:
y_gpt_lt("Расскажи про Москву")

'Москва — столица Российской Федерации и крупнейший город страны. Это один из важнейших политических, экономических, культурных и научных центров мира.\n\nГород расположен на реке Москве в европейской части России. В нём проживает более 13 миллионов человек, что делает его одним из самых густонаселённых городов Европы.\n\nМосква является историческим городом с богатой историей и культурой. Здесь расположено множество достопримечательностей, таких как Кремль, Красная площадь, Третьяковская галерея, Большой театр и другие.\n\nВ городе хорошо развита инфраструктура: здесь есть метро, автобусы, трамваи, троллейбусы, такси и каршеринг. Также в Москве расположены крупные транспортные узлы: три международных аэропорта, девять железнодорожных вокзалов, многочисленные станции пригородных поездов и две станции метрополитена.\n\nЭкономика Москвы является одной из крупнейших в стране и мире. Город является центром промышленности, финансов, торговли и услуг. Здесь расположены офисы крупных компаний

Москва — столица Российской Федерации и крупнейший город страны. Это один из важнейших политических, экономических, культурных и научных центров мира.

Город расположен на реке Москве в европейской части России. В нём проживает более 13 миллионов человек, что делает его одним из самых густонаселённых городов Европы.

Москва является историческим городом с богатой историей и культурой. Здесь расположено множество достопримечательностей, таких как Кремль, Красная площадь, Третьяковская галерея, Большой театр и другие.

В городе хорошо развита инфраструктура: здесь есть метро, автобусы, трамваи, троллейбусы, такси и каршеринг. Также в Москве расположены крупные транспортные узлы: три международных аэропорта, девять железнодорожных вокзалов, многочисленные станции пригородных поездов и две станции метрополитена.

Экономика Москвы является одной из крупнейших в стране и мире. Город является центром промышленности, финансов, торговли и услуг. Здесь расположены офисы крупных компаний, банки, торговые центры и многое другое.

Образование в Москве также находится на высоком уровне. В городе расположены ведущие университеты и научные учреждения страны, такие как Московский государственный университет имени М.В. Ломоносова, Высшая школа экономики и многие другие.

Кроме того, Москва является популярным туристическим направлением. Ежегодно её посещают миллионы туристов со всего мира, чтобы увидеть её достопримечательности, посетить музеи, театры и выставки.

Это лишь некоторые аспекты жизни в Москве. Город продолжает развиваться и меняться, предлагая новые возможности для жителей и гостей.

In [ ]:
y_gpt_rc4 = YandexGPT(model_uri="gpt://b1gt3tgpol6rntep2m65/yandexgpt/rc")

In [ ]:
y_gpt_rc4("Расскажи про Москву")

'Москва — столица России, один из крупнейших городов мира. Город расположен в центре европейской части страны на реке Москве. Население Москвы составляет более 13 миллионов человек.\n\nМосква является политическим, экономическим и культурным центром России. Здесь расположены федеральные органы государственнойской власти, крупные банки и компании, а также многочисленные музеи, театры и другие культурные учреждения.\n\nВ Москве можно найти множество достопримечательностей, которые привлекают туристов со всего мира. Среди них:\n* Московский Кремль и Красная площадь;\n* Храм Христа Спасителя;\n* Новодевичий монастырь;\n* Арбат;\n* ВДНХ (Выставка достижений народного хозяйства);\n* Парк Горького;\n* Третьяковская галерея;\n* Большой театр;\n* Исторический музей.\n\nГород также известен своими парками и зонами отдыха, где можно насладиться красивыми пейзажами и свежим воздухом. В Москве проходят различные фестивали, концерты и мероприятия, которые делают жизнь в городе интересной и разнообра

Москва — столица России, один из крупнейших городов мира. Город расположен в центре европейской части страны на реке Москве. Население Москвы составляет более 13 миллионов человек.

Москва является политическим, экономическим и культурным центром России. Здесь расположены федеральные органы государственнойской власти, крупные банки и компании, а также многочисленные музеи, театры и другие культурные учреждения.

В Москве можно найти множество достопримечательностей, которые привлекают туристов со всего мира. Среди них:
* Московский Кремль и Красная площадь;
* Храм Христа Спасителя;
* Новодевичий монастырь;
* Арбат;
* ВДНХ (Выставка достижений народного хозяйства);
* Парк Горького;
* Третьяковская галерея;
* Большой театр;
* Исторический музей.

Город также известен своими парками и зонами отдыха, где можно насладиться красивыми пейзажами и свежим воздухом. В Москве проходят различные фестивали, концерты и мероприятия, которые делают жизнь в городе интересной и разнообразной.

Транспортная система Москвы включает в себя метро, автобусы, трамваи и такси. Метро является одним из самых быстрых и удобных способов передвижения по городу. Также в Москве есть несколько аэропортов, откуда можно добраться до других городов России и мира.

In [ ]:
y_gpt_rc4("Расскажи про Москву подробно. Расскажи так чтобы захотелось приехать!")

'Москва — столица и крупнейший город России, расположенный в центре европейской части страны. Это один из самых больших и красивых городов мира, который привлекает туристов со всего света своими историческими достопримечательностями, богатой культурой и разнообразными развлечениями.\n\n**Исторические достопримечательности:**\n\nМосква — это город с богатой историей, которая отражается в его архитектуре. Здесь можно увидеть множество исторических зданий и сооружений, которые являются настоящими шедеврами архитектуры. Среди них — Московский Кремль, Красная площадь, храм Василия Блаженного, Новодевичий монастырь, Коломенское и другие.\n\nМосковский Кремль — это символ России и один из самых узнаваемых архитектурных комплексов в мире. Он расположен на берегу Москвы-реки и является резиденцией президента Российской Федерации. На территории Кремля можно увидеть Успенский собор, Архангельский собор, Благовещенский собор, Грановитую палату и другие исторические здания.\n\nКрасная площадь — это

Москва — столица и крупнейший город России, расположенный в центре европейской части страны. Это один из самых больших и красивых городов мира, который привлекает туристов со всего света своими историческими достопримечательностями, богатой культурой и разнообразными развлечениями.

**Исторические достопримечательности:**

Москва — это город с богатой историей, которая отражается в его архитектуре. Здесь можно увидеть множество исторических зданий и сооружений, которые являются настоящими шедеврами архитектуры. Среди них — Московский Кремль, Красная площадь, храм Василия Блаженного, Новодевичий монастырь, Коломенское и другие.

Московский Кремль — это символ России и один из самых узнаваемых архитектурных комплексов в мире. Он расположен на берегу Москвы-реки и является резиденцией президента Российской Федерации. На территории Кремля можно увидеть Успенский собор, Архангельский собор, Благовещенский собор, Грановитую палату и другие исторические здания.

Красная площадь — это главная площадь Москвы, которая является одним из самых известных мест в России. Здесь проходят парады, концерты и другие мероприятия. На Красной площади расположены такие известные достопримечательности, как Мавзолей Ленина, Собор Василия Блаженного и памятник Минину и Пожарскому.

**Культурная жизнь:**

Москва предлагает своим гостям богатую культурную программу. Здесь можно посетить музеи, театры, концертные залы и другие культурные учреждения. В Москве есть много интересных музеев, таких как Государственный исторический музей, Третьяковская галерея, Музей изобразительных искусств имени А. С. Пушкина и другие.

Театры Москвы предлагают широкий выбор спектаклей, от классических до современных. В городе также проходит множество концертов, фестивалей и других мероприятий.

**Развлечения:**

В Москве можно найти развлечения на любой вкус. Для любителей активного отдыха есть множество парков, где можно покататься на велосипедах, роликах или просто прогуляться. Также в Москве есть аквапарки, зоопарки и другие развлекательные центры.

Для тех, кто интересуется ночной жизнью, в Москве есть множество клубов, баров и ресторанов. Здесь можно попробовать блюда разных кухонь мира и насладиться атмосферой ночного города.

Кроме того, Москва — это большой шопинг-центр. Здесь есть множество магазинов, торговых центров и рынков, где можно купить всё, что угодно.

Если вы хотите узнать больше о Москве, посетите её официальные сайты и страницы в социальных сетях. Вы сможете найти информацию о достопримечательностях, культурных мероприятиях, развлечениях и других интересных местах. И, конечно же, не забудьте взять с собой фотоаппарат или камеру, чтобы запечатлеть свои впечатления от этого прекрасного города!

### Chat models ChatYandexGPT example

In [ ]:
from langchain_community.chat_models import ChatYandexGPT
from langchain_core.messages import HumanMessage, SystemMessage

In [ ]:
chat_model = ChatYandexGPT()

In [ ]:
answer = chat_model.invoke(
    [
        SystemMessage(
            content="You are a helpful assistant that translates English to French."
        ),
        HumanMessage(content="I love programming."),
    ]
)
answer

AIMessage(content='J’adore programmer.', additional_kwargs={}, response_metadata={}, id='run-02ade3b2-2a12-4355-bbca-d84855e2747b-0')

### Embedding models YandexGPT example

In [ ]:
from langchain_community.embeddings.yandex import YandexGPTEmbeddings

In [ ]:
embeddings = YandexGPTEmbeddings()

In [ ]:
text = "This is a test document."

In [ ]:
query_result = embeddings.embed_query(text)

In [ ]:
query_result[:5]

[-0.0214080810546875,
 0.096435546875,
 -0.04693603515625,
 -0.0183258056640625,
 -0.00551605224609375]

In [ ]:
query_result[:20]

[-0.0214080810546875,
 0.096435546875,
 -0.04693603515625,
 -0.0183258056640625,
 -0.00551605224609375,
 0.005657196044921875,
 -0.0931396484375,
 -0.0623779296875,
 0.09149169921875,
 0.0167999267578125,
 0.05517578125,
 -0.09722900390625,
 0.0263214111328125,
 -0.0933837890625,
 -0.05670166015625,
 0.06170654296875,
 0.031982421875,
 -0.055816650390625,
 -0.11517333984375,
 -0.056610107421875]

In [ ]:
doc_result = embeddings.embed_documents([text])

In [ ]:
doc_result[0][:20]

[-0.0074462890625,
 0.055450439453125,
 -0.0195159912109375,
 -0.01561737060546875,
 0.04034423828125,
 0.003391265869140625,
 0.01045989990234375,
 -0.0247650146484375,
 0.037445068359375,
 -0.053924560546875,
 0.056060791015625,
 -0.0950927734375,
 -0.01715087890625,
 -0.0657958984375,
 -0.06103515625,
 0.09149169921875,
 -0.0047454833984375,
 -0.062225341796875,
 -0.07415771484375,
 0.0047607421875]

### Parser YandexSTTParser

In [ ]:
# from langchain_community.document_loaders import YandexSTTParser